apps/bot/src/services/twitch-eventsub/index.ts

```diff
@@ -0,0 +1,131 @@
+import { logger } from "utils/logger";
+import { type MessageEvent, WebSocket } from "ws";
+import { SEventSubGeneralMessage, SEventSubMetadata } from "./types";
+import { handleEvents } from "./events";
+
+export class TwitchEventSub {
+  private ws: WebSocket;
+
+  private static _instance: TwitchEventSub | undefined;
+
+  private constructor(private url: string) {
+    this.ws = new WebSocket(url);
+  }
+
+  public static instance(url: string) {
+    if (!TwitchEventSub._instance) {
+      TwitchEventSub._instance = new TwitchEventSub(url);
+      return TwitchEventSub._instance;
+    }
+
+    return TwitchEventSub._instance;
+  }
+
+  public async connect(): Promise<void> {
+    await new Promise((resolve) => {
+      try {
+        this.ws.addEventListener("open", () => {
+          console.log("connection is open");
+        });
+        this.ws.addEventListener("message", (r) => this.onMessage(r));
+      } catch (_) {
+        resolve(undefined);
+      }
+    });
+  }
+
+  private handlePingMessage(msg: string): void {
+    if (msg.startsWith("PING")) {
+      const pong = `PONG${msg.substring("PING".length)}`;
+      this.ws.send(pong);
+    }
+  }
+
+  private keepaliveTimeout = 10;
+  private lastevent = 0;
```

Comment:
It's just for organization, but usually, it's worth it to keep all the internal `data` of the class on top of it, to keep it easy to look in case of need.

Ps.: If you feel that you need the variables close to a specific function cuz it will be only used in this specific function, it's is an indication that you need to abstract this values into it's own class/object